Manual block file pruning.

mrbandrews opened the pull request:
Implements #7365.

Now there is auto-prune and manual-prune.  The user enables manual pruning on the command line with prune=1 and then uses an RPC command to prune:  "pruneblockchain X" to prune up to height X. 

Updated the python test (pruning.py).

jonasschnelli commented:
Nice!
Concept ACK.

instagibbs commented:
What use cases are motivating this mode?

laanwj commented:
Concept ACK

> What use cases are motivating this mode?

This is explained in the original issue, #7365: if you have another application that needs to consume the block data before it's gone.

instagibbs commented:
Should have read the issue, sorry. Concept ACK

kazcw commented:
I'm wondering if the command line option is necessary. I see a few functions it serves:
1. acts as a check on the RPC call? This is unnecessary (trust the authenticated RPC user)
2. tells bitcoind it's okay that blockfiles are missing. This is covered by the persistent flag that's read into fHavePruned.
3. Prevents the combination of prune automatically + manually. While I'm not sure of the use case for both, the combination has a straightforward interpretation. Why go out of our way to prevent it?
Anything I'm missing?

kazcw commented:
4. There's no way to tell our existing peers our nServices have changed when the RPC call comes in...

sipa commented:
Concept ACK. Needs rebase.

mrbandrews commented:
Rebased.

rebroad commented on src/main.cpp:
```diff
@@ -3617,10 +3618,44 @@ void UnlinkPrunedFiles(std::set<int>& setFilesToPrune)
+/* This function is called from the RPC code for pruneblockchain */
+void PruneBlockFilesManual(int manualPruneHeight)
+{
+    // stuff the prune height into a global variable, and flush
```
I usually get told off for using global variables. Curious to know when they are permitted and when they are discouraged.

sipa commented:
Needs rebase.

luke-jr commented on src/main.cpp:
```diff
@@ -3842,10 +3843,44 @@ void UnlinkPrunedFiles(std::set<int>& setFilesToPrune)
 {
     LOCK2(cs_main, cs_LastBlockFile);
+    const CChainParams& chainparams = Params();
+    uint64_t nPruneAfterHeight = chainparams.PruneAfterHeight();
```
We're trying to reduce calls to global Params() by passing things as arguments. This change has the opposite effect for no clear reason.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -670,6 +670,32 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+
+    int height = params[0].get_int();
+    if (height < 0) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Negative block height.");
```
`RPC_INVALID_PARAMETER` seems more appropriate here.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -670,6 +670,32 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Negative block height.");
+    } else if ((unsigned int) chainActive.Height() < Params().PruneAfterHeight()) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Blockchain is too short for pruning.");
+    }
```
Should this also check if the requested height is too close to the tip? Or maybe just return the height it was able to successfully prune up to...

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -670,6 +670,32 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+            "1. \"height\"       (int, required) The block height to prune up to.\n");
+
+    if (pruneMode != PRUNE_MANUAL) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Cannot prune via RPC unless in manual prune mode.");
```
`RPC_METHOD_NOT_FOUND` seems better for this (already used for wallet RPCs when the wallet is not enabled)

luke-jr commented:
Oh, it might make sense to add "prunemode": "auto"/"manual" to `getblockchaininfo` also.

petertodd commented:
Concept ACK

My OpenTimestamps Server could use this!

mrbandrews commented:
Rebased and feedback addressed.  I didn't make a couple of the suggested changes, though:
1) "int" for block height is used elsewhere in the code (chain.h/chain.cpp); perhaps a future PR could change the type used in each place.
2) Params() - this was just a code move, although I had left one unnecessary line of code in there, which I removed now.
3) heights close to tip - it will prune up to the limit and log.

laanwj commented on src/main.h:
```diff
@@ -191,8 +191,10 @@ static const uint64_t nMinDiskSpace = 52428800;
 extern bool fHavePruned;
-/** True if we're running in -prune mode. */
-extern bool fPruneMode;
+enum PruneMode {PRUNE_NONE = 0, PRUNE_AUTO, PRUNE_MANUAL };
```
Let's use C++11 scoped enums in new code
```
enum struct PruneMode { NONE=0, AUTO, MANUAL };
```
Then refer to `PruneMode::NONE` etc.

laanwj commented on src/init.cpp:
```diff
@@ -928,12 +928,15 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     }
     nPruneTarget = (uint64_t) nSignedPruneTarget;
-    if (nPruneTarget) {
+    if (nPruneTarget / 1024 / 1024 == 1) {  // manual pruning: -prune=1
```
Can we please store and compare the option value before multiplication? Doing a division here, though it achieves the correct effect, seems circuitous and unclear.

laanwj commented on src/main.cpp:
```diff
@@ -3317,7 +3320,7 @@ bool FindBlockPos(CValidationState &state, CDiskBlockPos &pos, unsigned int nAdd
         unsigned int nNewChunks = (vinfoBlockFile[nFile].nSize + BLOCKFILE_CHUNK_SIZE - 1) / BLOCKFILE_CHUNK_SIZE;
         if (nNewChunks > nOldChunks) {
-            if (fPruneMode)
+            if (pruneMode == PRUNE_AUTO)
```
Are you sure this only needs to be done in auto-pruning mode?

laanwj commented on src/wallet/rpcdump.cpp:
```diff
@@ -115,7 +115,7 @@ UniValue importprivkey(const JSONRPCRequest& request)
         fRescan = request.params[2].get_bool();
 
-    if (fRescan && fPruneMode)
+    if (fRescan && pruneMode)
```
As you already have to change every line on which the variable occurs anyhow, I'd prefer explicitly comparing the enumeration against a value instead of using it like a boolean: e.g. `pruneMode != PruneMode::NONE`. This makes it  clear to developers reading the code that this is an enumeration and not a boolean and can avoid them introducing silly bugs.

laanwj commented on src/main.cpp:
```diff
@@ -3851,10 +3854,40 @@ void UnlinkPrunedFiles(std::set<int>& setFilesToPrune)
+void FindFilesToPruneManual(std::set<int>& setFilesToPrune, int nManualPruneHeight)
+{
+    LOCK2(cs_main, cs_LastBlockFile);
+    if (chainActive.Tip() == NULL || pruneMode != PRUNE_MANUAL) {
```
Calling this function with pruneMode != PRUNE_MANUAL should be an assertion error, as it must be a bug in the code.

laanwj commented on src/rpc/blockchain.cpp:
```diff
@@ -814,6 +814,32 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+
+    if (pruneMode != PRUNE_MANUAL) {
+        throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Cannot prune via RPC unless in manual prune mode.");
+        return 0;
```
No need for a return if you have a throw

laanwj commented on src/rpc/blockchain.cpp:
```diff
@@ -814,6 +814,32 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Negative block height.");
+    } else if ((unsigned int) chainActive.Height() < Params().PruneAfterHeight()) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Blockchain is too short for pruning.");
+    }
```
Do we need any foot-shooting checks: w.g. that the last 144 blocks are being retained to be robust against reorgs?

mrbandrews commented:
Feedback addressed.  Re: the code in FindBlockPos (which reset fCheckForPruning only in auto-prune mode), yes that should only be auto-prune.  I designed manual pruning to be a one-time act of pruning to the specified height.  (In looking at this code, though, I noticed that nManualPruneHeight should probably be set=0 after we do the manual pruning, so I added that line of code, in a separate commit.) 

So, there's a commit responding to laanwj's feedback, and another with that one line of code.

mrbandrews commented:
There was a merge conflict (the declaration of FlushStateToDisk near the top of main.cpp conflicted with my adding an optional parameter), so I rebased and squashed everything.  This should work now.

sipa commented on src/main.cpp:
```diff
@@ -4140,7 +4175,7 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
         if (pindex->nHeight < chainActive.Height()-nCheckDepth)
             break;
-        if (fPruneMode && !(pindex->nStatus & BLOCK_HAVE_DATA)) {
+        if (pruneMode != PruneMode::NONE && !(pindex->nStatus & BLOCK_HAVE_DATA)) {
```
`FindFilesToPruneAuto` is only called when `pruneMode == PruneMode::AUTO`, so why do we need this test? Maybe turn it into an assert at the beginning of the function.

sipa replied:
Also, `if (pruneMode)` is equivalent to `if (pruneMode != PruneMode::NONE)`, since PruneMode::NONE is explicitly defined as 0.

(applies to many changed lines in this PR)

sipa commented on src/main.h:
```diff
@@ -273,7 +280,8 @@ CAmount GetBlockSubsidy(int nHeight, const Consensus::Params& consensusParams);
  */
-void FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPruneAfterHeight);
+void FindFilesToPruneAuto(std::set<int>& setFilesToPrune);
+void FindFilesToPruneManual(std::set<int>& setFilesToPrune, int nManualPruneHeight);
```
There are many functions in main that are not exported (see the whole namespace block for the handling of NodeState, for example).

sipa commented on src/main.h:
```diff
@@ -191,8 +191,10 @@ static const uint64_t nMinDiskSpace = 52428800;
 extern bool fHavePruned;
-/** True if we're running in -prune mode. */
-extern bool fPruneMode;
+enum struct PruneMode {NONE=0, AUTO, MANUAL};
```
Meta question: why do we need a tristate here? I think we could allow manual pruning even when in 'auto` mode. In that case, manual-only pruning could be requested by setting the limit very high.

mrbandrews committed:
Add pruneblockchain RPC to enable manual block file pruning.

mrbandrews commented:
So, originally I thought it was simpler to make manual pruning a separate mode from autoprune.  After sipa's question and giving this a little more thought, I agree that it should be the same.  So, now setting prune=1 (manual pruning) means autoprune with target=max. 

Sorry for the late change in approach.  This makes the diff smaller, though. 

Also, now I only expose in main.h that which is used elsewhere.

mrbandrews commented:
Rebased.

ryanofsky commented on qa/rpc-tests/pruning.py:
```diff
@@ -212,6 +227,90 @@ def reorg_back(self):
+
+        # should not prune because chain tip of node 3 (995) < PruneAfterHeight (1000)
+        try:
+            self.nodes[3].pruneblockchain(500)
```
Might be good to check for the specific error messages here and below, e.g. `assert_raises_message(JSONRPCException, "Blockchain is too short for pruning.", self.nodes[3].pruneblockchain, 500)`

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -814,6 +814,36 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+            "1. \"height\"       (int, required) The block height to prune up to.\n");
+
+    if (!fPruneMode)
+        throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Cannot prune blocks because node is not in prune mode.");
```
Could add a python test for this condition.

ryanofsky reviewed (approved):
Lightly tested ACK 485684c748b477caee06bcaad207eeb805b6c3a2 (just ran pruning.py).

Code looks really good. I made a few suggestions you could consider, but they're all on the python test.

mrbandrews commented:
Added a commit with edits to python test per ryanofsky review.

neocogent commented:
This is very useful for me. I'm doing some testing on this now. My application is a sql db on top of the blockchain that depends on processing incoming blocks. After processed into sql the block data is never used again so it's ideal if it can be pruned to save space, especially as I'm paying per GB on a cloud server.

Manual pruning allows me to start processing blocks while still syncing the chain without worry about blocks being pruned before I get to them. I had a workaround for this that disabled network access when the pruning was approaching the current block being worked on. So this is a much nicer approach and I feel there are probably more use cases for this. 

I noticed that the man pages are not updated with this. The cmd line help does indicate prune=1 available but the man page misses that updated detail. So it may be nice to update that page too.

Currently it appears to be working but I am still in progress with a new full sync and I'm watching that pruning of block files trails behind my block processor as expected.

ryanofsky commented:
I rebased this, updated the man pages, and squashed the test commit.

gmaxwell commented on src/rpc/blockchain.cpp:
```diff
@@ -814,6 +814,36 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+    unsigned int height = (unsigned int) heightParam;
+    unsigned int chainHeight = (unsigned int) chainActive.Height();
+    if (chainHeight < Params().PruneAfterHeight())
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Blockchain is too short for pruning.");
```
This error is a bit confusing. "Blockchain is already pruned past the requested height." might be better. Or even just silently being successful, since it is already pruned that far. :)

gmaxwell commented:
utACK. Great feature!

gmaxwell commented:
One question: I view this feature as the perfect compliment to importmulti, that lets you be sure you've imported all your keys before you prune.  But importmulti takes its scanning argument as a timestamp, while this takes it's argument as a height.  Should we also support a timestamp option here that uses the same criteria as import multi?

ryanofsky committed:
fixup! Add pruneblockchain RPC to enable manual block file pruning.

Extend pruneblockchain RPC to accept block timestamps as well as block indices.

ryanofsky commented:
Rebased for named arguments.

@gmaxwell, I extended pruneblockchain in afffeea7d98ba358acd54a89bc0e7ae1c4d54023 to be able to take a timestamp instead of a block index. The code change is pretty small. The test change is bigger but straightforward.

laanwj commented:
utACK afffeea

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -814,6 +814,46 @@ static bool GetUTXOStats(CCoinsView *view, CCoinsStats &stats)
+    else if (height > chainHeight)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Blockchain is shorter than the attempted prune height.");
+    else if (height > chainHeight - MIN_BLOCKS_TO_KEEP)
+        LogPrint("rpc", "Attempt to prune blocks close to the tip.  Retaining the minimum number of blocks.");
```
one post merge nit:
We should probably report the pruned height in case the user given value was overrode.

ryanofsky replied:
Done in https://github.com/bitcoin/bitcoin/pull/9518
